accord-core/src/test/java/accord/messages/ReadDataTest.java

```diff
@@ -0,0 +1,159 @@
+        Ranges ranges = Ranges.single(IntKey.range(0, 1000));
+        IntKey.Raw key = IntKey.key(42);
+        Keys keys = Keys.of(key);
+
+        AsyncResults.SettableResult<Data> readResult = new AsyncResults.SettableResult<>();
+        Data data = Mockito.mock(Data.class);
+
+        Read read = Mockito.mock(Read.class);
+        Mockito.when(read.slice(any())).thenReturn(read);
+        Mockito.when(read.merge(any())).thenReturn(read);
+        Mockito.when(read.read(any(), any(), any(), any(), any())).thenAnswer(new Answer<AsyncChain<Data>>()
+        {
+            private boolean called = false;
+            @Override
+            public AsyncChain<Data> answer(InvocationOnMock ignore) throws Throwable
+            {
+                if (called) throw new IllegalStateException("Multiple calls");
+                return readResult;
+            }
+        });
+        Query query = Mockito.mock(Query.class);
+        Update update = Mockito.mock(Update.class);
+        Mockito.when(update.slice(any())).thenReturn(update);
+        Result result = Mockito.mock(Result.class);
+
+        Txn txn = new Txn.InMemory(keys, read, query, update);
+        PartialTxn partialTxn = txn.slice(ranges, true);
+        FullRoute<?> route = keys.toRoute(key.toUnseekable());
+        PartialRoute<?> partialRoute = route.slice(ranges);
+
+        //TODO ProgressKey must equal HomeKey to get Home shard; home key = key
+        RoutingKey progressKey = key.toUnseekable();
+
+        PartialDeps deps = PartialDeps.builder(ranges).build();
+        Timestamp executeAt = txnId;
+
+        AsyncChain<Void> writeResult = new AsyncResults.SettableResult<>();
+        Write write = Mockito.mock(Write.class);
+        Mockito.when(write.apply(any(), any(), any(), any())).thenReturn(writeResult);
+        Writes writes = new Writes(executeAt, keys, write);
+
+
+        CommandStore store = node.commandStores().unsafeForKey(key);
+        check(store.execute(PreLoadContext.contextFor(txnId, keys), safe -> {
+            CheckedCommands.preaccept(safe, txnId, partialTxn, route, progressKey);
+            CheckedCommands.accept(safe, txnId, Ballot.ZERO, partialRoute, partialTxn.keys(), progressKey, executeAt, deps);
+            CheckedCommands.commit(safe, txnId, route, progressKey, partialTxn, executeAt, deps);
+        }));
+
+        ReadData readData = new ReadData(node.id(), TOPOLOGIES, txnId, keys, txnId);
+        readData.process(node, node.id(), replyContext);
+
+        Mockito.verifyNoInteractions(sink);
+
+        check(store.execute(PreLoadContext.contextFor(txnId, keys), safe -> {
+            CheckedCommands.apply(safe, txnId, safe.latestEpoch(), route, executeAt, deps, writes, result);
+        }));
+
+        readResult.setSuccess(data);
+
+        Mockito.verify(sink, Mockito.atMost(2)).reply(Mockito.eq(node.id()), Mockito.eq(replyContext), Mockito.eq(ReadData.ReadNack.Redundant));
```

Comment:
`accord.messages.ReadData#apply` sets `isObsolete=true` without sending the nack, so there could be the following setup

store1: status=ReadyToExecute, send read
store2: status=PreApplied, set isObsolete

in this case when the read from `store1` completes it needs to know to ignore and that the nack wasn't set... I could switch this to an enum `{ALL_GOOD, SENT_NACK, NEED_TO_SEND_NACK}` which would avoid the second nack...  figured this was simplest

Reply:
actually, in that case `apply` sets the field but returns `Redundant` so `accord.messages.ReadData#accept` should send that...

and `accord.messages.ReadData#obsolete` sends directly... so under all cases we send?  ill play around with this

Reply:
fixed this and pushed tests to hit all 3 cases